src/test/java/bio/terra/workspace/service/resource/controlled/ControlledResourceServiceTest.java

```diff
@@ -252,32 +248,138 @@ public void createAiNotebookInstanceUndo() throws Exception {
+    Map<String, StepStatus> retrySteps = new HashMap<>();
+    retrySteps.put(
+        RetrieveNotebookServiceAccountStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    retrySteps.put(
+        DeleteAiNotebookInstanceStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    retrySteps.put(DeleteServiceAccountStep.class.getName(), StepStatus.STEP_RESULT_FAILURE_RETRY);
+    jobService.setFlightDebugInfoForTest(
+        FlightDebugInfo.newBuilder().doStepFailures(retrySteps).build());
+
+    controlledResourceService.deleteControlledResourceSync(
+        resource.getWorkspaceId(),
+        resource.getResourceId(),
+        userAccessUtils.defaultUserAuthRequest());
+    assertNotFound(instanceName, notebooks);
+    assertNotFound(serviceAccountName, crlService.getIamCow());
+    assertThrows(
+        ResourceNotFoundException.class,
+        () ->
+            controlledResourceService.getControlledResource(
+                resource.getWorkspaceId(),
+                resource.getResourceId(),
+                userAccessUtils.defaultUserAuthRequest()));
+  }
+
+  @Test
+  @DisabledIfEnvironmentVariable(named = "TEST_ENV", matches = bufferServiceDisabledEnvsRegEx)
+  public void deleteAiNotebookInstanceUndoIsDismalFailure() throws Exception {
+    ControlledAiNotebookInstanceResource resource =
+        createDefaultPrivateAiNotebookInstance("delete-ai-notebook-instance-undo");
+
+    // Test that trying to undo a notebook deletion is a dismal failure. We cannot undo deletion.
+    jobService.setFlightDebugInfoForTest(
+        FlightDebugInfo.newBuilder().lastStepFailure(true).build());
+    assertThrows(
+        InvalidResultStateException.class,
+        () ->
+            controlledResourceService.deleteControlledResourceSync(
+                resource.getWorkspaceId(),
+                resource.getResourceId(),
+                userAccessUtils.defaultUserAuthRequest()));
+  }
+
+  /** Create a controlled AI Notebook instance with default private settings. */
+  private ControlledAiNotebookInstanceResource createDefaultPrivateAiNotebookInstance(
+      String instanceId) {
+    Workspace workspace = reusableWorkspace();
+    ApiGcpAiNotebookInstanceCreationParameters creationParameters =
+        ControlledResourceFixtures.defaultNotebookCreationParameters()
+            .instanceId(instanceId)
+            .location(DEFAULT_NOTEBOOK_LOCATION);
+    ControlledAiNotebookInstanceResource.Builder resourceBuilder =
+        ControlledResourceFixtures.makeDefaultAiNotebookInstance()
+            .name(instanceId)
+            .workspaceId(workspace.getWorkspaceId())
+            .assignedUser(userAccessUtils.getDefaultUserEmail())
+            .accessScope(AccessScopeType.ACCESS_SCOPE_PRIVATE)
+            .managedBy(ManagedByType.MANAGED_BY_USER)
+            .instanceId(instanceId)
+            .location(DEFAULT_NOTEBOOK_LOCATION);
+    ControlledAiNotebookInstanceResource resource = resourceBuilder.build();
```

Comment:
nit: why not inline the builder declaration?

Reply:
done